xbmc/cores/AudioEngine/Sinks/AESinkXAudio.cpp

```diff
@@ -573,19 +654,33 @@ bool CAESinkXAudio::InitializeInternal(std::string deviceId, AEAudioFormat &form
   /* Set up returned sink format for engine */
   if (format.m_dataFormat != AE_FMT_RAW)
   {
-    if (wfxex.Format.wBitsPerSample == 32)
+    switch (wfxex.Format.wBitsPerSample)
     {
-      if (wfxex.SubFormat == KSDATAFORMAT_SUBTYPE_IEEE_FLOAT)
-        format.m_dataFormat = AE_FMT_FLOAT;
-      else if (wfxex.Samples.wValidBitsPerSample == 32)
-        format.m_dataFormat = AE_FMT_S32NE;
-      else
-        format.m_dataFormat = AE_FMT_S24NE4MSB;
+      case 32:
+        if (wfxex.SubFormat == KSDATAFORMAT_SUBTYPE_IEEE_FLOAT)
+          format.m_dataFormat = AE_FMT_FLOAT;
+        else if (wfxex.Samples.wValidBitsPerSample == 32)
+          format.m_dataFormat = AE_FMT_S32NE;
+        else
+          format.m_dataFormat = AE_FMT_S24NE4MSB;
+        break;
+
+      case 24:
+        format.m_dataFormat = AE_FMT_S24NE3;
+        break;
+
+      case 16:
+        format.m_dataFormat = AE_FMT_S16NE;
+        break;
+
+      case 8:
+        format.m_dataFormat = AE_FMT_U8;
+        break;
+
+      default:
+        format.m_dataFormat = AE_FMT_S16NE;
+        break;
```

Comment:
@fritsch wonder if that's right - this is the same behavior as before for an unsupported number of bits, but doesn't seem right, would you agree?
Being in the default case means that the OS/audio device accepted the format but the sink doesn't fully support it.
I think that the sink should either return an error, or test the number of bits of the AE provided format earlier and when it's not a recognized number, go into the loop that uses the testFormats list to come up with a format supported by the sink. Latter option is almost as easy and seems more forward looking to me.

Reply:
So in general during initializing we enumerate our supported formats in the sink and return those to AE. AE should then only call us with one of those after all.

The ordering is also important. Back the AE chose simply what was first in list.

On the next open though you can decide again, when overwriting the referenced format back to AE. Best is saving the supported list, updating on hot plug and using the biggest format available.

Pulse uses float, ALSA 32 bit int and android depending on multi channel or not 16 bit int or float.

Reason you can overwrite is also: sinks may not support all formats in all channel combinations. See android example.

Reply:
My testing and pretty much every log I come across show AE calling Initialize() with AE_FMT_FLOAT even when it was not added to the device during enumeration by the sink,
The WASAPI sink then always falls back to the predefined formats of testFormats to find and return a working alternative to AE. Which bypasses any smarts AE might have about format selection.
DSound and XAudio sinks enumerate and support float (the system mix format) so that behavior is not visible for them.

I added a default case for completeness but started to wonder what to do because current behavior doesn't seem right.

Reply:
I now remember.

There is not much smartness needed, reason: ffmpeg decoders produce float (planar) 32 bit. So AE tries that (non planar) directly, but sink can decide to opt out. So AE will then transform the format.

For passthrough it is checked if sink supports stream type raw and int16.

That way a platform sink can overrule any potential AE smartness and logic inside AE is totally simple.

Reply:
OK. In light of that information I'll withdraw that commit as the 8 bit format doesn't make much sense either.